src/undefined_behavior.md

```diff
@@ -0,0 +1,145 @@
+it's often tempting to think of undefined behavior as telling your program "if
+you get here, do whatever you want". However, undefined behavior is a purely
+compile-time concept. It's not telling your program "do whatever you want", it's
+telling the compiler "assume this can never happen". The compiler may not have a
+better response than saying "if you get here, panic". Or, it may be able to use
+that promise to better optimize your code.
+
+As an example, consider this Rust code:
+
+```rust
+use std::hint::unreachable_unchecked;
+
+unsafe fn char_to_int(c: char) -> u8 {
+    match c {
+        '0' => 0,
+        '1' => 1,
+        '2' => 2,
+        ...
+        '9' => 9,
+        _ => unsafe { unreachable_unchecked() },
+    }
+}
+```
+
+This converts a char `'0'..'9'` to its corresponding integer value. In the last
+arm of the match, we call `unreachable_unchecked()`, which is a compiler hint
+that says "it would be undefined behavior to reach here". Because we promised
+the compiler that `c` won't be any value other than `0..9`, it could optimize
+this function into something like:
+
+```rust
+unsafe fn char_to_int(c: char) -> u8 {
+    c as u8 - '0' as u8
+}
+```
+
+Consider what would happen if we called `char_to_int('A')`. `'A'` has a value of
+`65` as a `u8`, and `'0'` has a value of `48` as a `u8`. So the optimized
+version of our function would return `17`. But what if the compiler chose to
+optimize our function a different way instead:
+
+```rust
+unsafe fn char_to_int(c: char) -> u8 {
+    c as u8 & 0b1111
+}
+```
+
+This does the same thing as our original version for all characters `'0'..'9'`.
+Now consider what _this version_ of our function would do if we called
+`char_to_int('A')`. `'A'` has a value of `65` (`0b0010_0001` in binary
+notation), so this version would return `0010_0111 & 0000_1111 = 0111 = 15`.
+This is a different result than we would have gotten with the previous
+optimization!
+
+Finally, consider this version:
+
+```rust
+unsafe fn char_to_int(c: char) -> u8 {
+    if c as u8 < '0' as u8 || c > '9' as u8 {
+        panic!()
+    } else {
+        c as u8 - '0' as u8
+    }
+}
+```
+
+This version doesn't return anything, it panics! In fact, the compiler would be
+allowed to put _anything_ in where the `panic!()` is located; the resulting
+function would be just as correct and optimal as this one.
+
+This strikes at the core of what "undefined behavior" is. The Rust compiler
+transforms code based on a set of assumptions that always hold. If you break one
+of these assumptions, then the behavior of your program is undefined because
+it's impossible to know what transforms the compiler is doing based on them.
+
+## Unsoundness
+
+Now that we know exactly what undefined behavior is, we can understand what it
+means for some Rust code to be _unsound_. Unsound code is just code that can
+trigger undefined behavior even if no mistakes are made.
```

Comment:
suggestion: we should be very explicit about the thing about unsound being both:

 - a concrete chunk of code definitely triggering UB
 - unsafe abstractions being able to trigger UB when called from safe code

Reply:
Did my best to incorporate that in, does what I wrote make sense?